Peek at queued task's latency in a thread pool

DiannaHohensee opened the pull request:
Relates ES-12233

-------
This was described [here](https://github.com/elastic/elasticsearch/pull/130373#discussion_r2178294244) and requested [here](https://github.com/elastic/elasticsearch/pull/130373#issuecomment-3053336546).

This has a lot of type casting.. I'm not sure what our policy is on using this much type casting in production code?

DiannaHohensee committed:
Peek at queued task's latency in a thread pool

elasticsearchmachine commented:
Pinging @elastic/es-distributed-coordination (Team:Distributed Coordination)

DiannaHohensee commented on server/src/test/java/org/elasticsearch/common/util/concurrent/TaskExecutionTimeTrackingEsThreadPoolExecutorTests.java:
```diff
@@ -93,14 +93,81 @@ public void testExecutionEWMACalculation() throws Exception {
+    public void testFrontOfQueueLatency() throws Exception {
         ThreadContext context = new ThreadContext(Settings.EMPTY);
-        RecordingMeterRegistry meterRegistry = new RecordingMeterRegistry();
-        final var threadPoolName = randomIdentifier();
```
Both of these were unused, so I took the opportunity to delete them. Does mess up the diff, a bit, though.

DiannaHohensee commented on server/src/test/java/org/elasticsearch/common/util/concurrent/TaskExecutionTimeTrackingEsThreadPoolExecutorTests.java:
```diff
@@ -93,14 +93,81 @@ public void testExecutionEWMACalculation() throws Exception {
+                // Release any potentially running task. This could be racy (a task may start executing and hit the barrier afterward) and
+                // is best-effort.
+                safeAwait(barrier);
+            }
```
The test failure output focuses on CyclicBarrier timeout because a Task is hanging during execution in the thread pool. An assert from the try-block is also reported if you dig around the failure stacks, but I think this might help make failure causes more obvious when a task is left running.

DiannaHohensee commented:
@nicktindall Not sure if you've been busy, or have concerns whether we want to continue using queue latency? I brought up a new discussion thread in the project channel to clarify. I do think queue latency is still important.

nicktindall reviewed (commented):
Sorry for the delay here, not sure what was discussed today, but let's hold off merging until we know what (if any) role queue latency will play in the decider.

DiannaHohensee committed:
Merge branch 'main' into 2025/07/15/peek-queue-head

DiannaHohensee committed:
peek at the queue twice, ensure the queue duration reported the second peek is longer than the first

DiannaHohensee commented on server/src/test/java/org/elasticsearch/common/util/concurrent/TaskExecutionTimeTrackingEsThreadPoolExecutorTests.java:
```diff
@@ -93,14 +93,81 @@ public void testExecutionEWMACalculation() throws Exception {
         final var barrier = new CyclicBarrier(2);
+        // Replace all tasks submitted to the thread pool with a configurable task that supports configuring queue latency durations and
+        // waiting for task execution to begin via the supplied barrier.
         var adjustableTimedRunnable = new AdjustableQueueTimeWithExecutionBarrierTimedRunnable(
```
See https://github.com/elastic/elasticsearch/pull/131329/commits/e3d7d0e798f39b90b002fea86de4403b6bfdd91e. I also added a second peek() call to the test to check that the queued duration reported does increase 👍 Good idea.

DiannaHohensee committed:
time unit change

DiannaHohensee commented on server/src/test/java/org/elasticsearch/common/util/concurrent/TaskExecutionTimeTrackingEsThreadPoolExecutorTests.java:
```diff
@@ -93,14 +93,81 @@ public void testExecutionEWMACalculation() throws Exception {
         var adjustableTimedRunnable = new AdjustableQueueTimeWithExecutionBarrierTimedRunnable(
             barrier,
-            TimeUnit.NANOSECONDS.toNanos(1000000)
+            TimeUnit.NANOSECONDS.toNanos(1000000) // Until changed, queue latencies will always be 1 millisecond.
```
Sure, changed https://github.com/elastic/elasticsearch/pull/131329/commits/20641afcbd9cc99a71f3f656d54c7234c38d467c

DiannaHohensee commented:
Thanks for the review. Yep, we'll discuss in the sync 👍

DaveCTurner reviewed (commented):
Makes sense to me

DiannaHohensee committed:
Merge branch 'main' into 2025/07/15/peek-queue-head

DiannaHohensee committed:
Removed best-effort while-loops
Replaces executor shutdowns with more reliable ThreadPool#terminate calls
Added assertBusy around queue latency checks, to avoid races with ThreadPool clock not moving forward

DiannaHohensee committed:
Merge branch 'main' into 2025/07/15/peek-queue-head

DiannaHohensee reviewed (commented):
Thanks for the review. I've updated per the feedback in https://github.com/elastic/elasticsearch/pull/131329/commits/44e5f18bf87af8c5f48882530538a1ce99f909ff

DiannaHohensee committed:
Merge branch 'main' into 2025/07/15/peek-queue-head

DiannaHohensee committed:
add method to ensure time passes, replace assertBusy

DiannaHohensee commented on server/src/test/java/org/elasticsearch/common/util/concurrent/TaskExecutionTimeTrackingEsThreadPoolExecutorTests.java:
```diff
@@ -93,14 +93,93 @@ public void testExecutionEWMACalculation() throws Exception {
+
+            var frontOfQueueDuration = executor.peekMaxQueueLatencyInQueue();
+            assertThat("Expected a task to be queued", frontOfQueueDuration, greaterThan(0L));
+            safeSleep(10);
```
Ah, that is simpler than I was initially imagining. Updated in https://github.com/elastic/elasticsearch/pull/131329/commits/7b294022da986e14d52e15c1ecf3be0f7a10fd6f, I think that's what you mean

DiannaHohensee reviewed (commented):
Replaced the assertBusy with a check to ensure time passes: https://github.com/elastic/elasticsearch/pull/131329/commits/7b294022da986e14d52e15c1ecf3be0f7a10fd6f

DiannaHohensee committed:
Merge branch 'main' into 2025/07/15/peek-queue-head

DaveCTurner commented on server/src/test/java/org/elasticsearch/common/util/concurrent/TaskExecutionTimeTrackingEsThreadPoolExecutorTests.java:
```diff
@@ -387,4 +455,15 @@ long getQueueTimeNanos() {
+     */
+    private static void waitForTimeToElapse() throws InterruptedException {
+        final var startNanoTime = System.nanoTime();
+        while (TimeUnit.MILLISECONDS.convert(System.nanoTime() - startNanoTime, TimeUnit.NANOSECONDS) <= 100) {
+            Thread.sleep(100);
```
I think the test will be fine if the clock advances at all, even by 1ns - no need to keep retrying until it hits 100,000,000ns :)

DiannaHohensee replied:
Updated to an itty-bitty single nano :) Not sure I'll ever write such a small number again.

DaveCTurner reviewed (approved):
LGTM

DiannaHohensee committed:
Merge branch 'main' into 2025/07/15/peek-queue-head

DiannaHohensee committed:
update sleep to be very small -- any amount of time is sufficient
